go/teams/box_audit.go

```diff
@@ -0,0 +1,1099 @@
+
+func (j *BoxAuditJail) GetVersion() Version {
+	return j.Version
+}
+
+func NewBoxAuditJail(version Version) *BoxAuditJail {
+	return &BoxAuditJail{
+		TeamIDs: make(map[keybase1.TeamID]bool),
+		Version: version,
+	}
+}
+
+func (a *BoxAuditor) shouldAudit(mctx libkb.MetaContext, team Team) (bool, *keybase1.BoxAuditAttemptResult, error) {
+	// if team.IsSubteam() {
+	// 	res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_SUBTEAM
+	// 	return false, &res, nil
+	// }
+	if team.IsOpen() {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_OPENTEAM
+		return false, &res, nil
+	}
+	role, err := team.MemberRole(mctx.Ctx(), mctx.CurrentUserVersion())
+	if err != nil {
+		return false, nil, err
+	}
+	if !role.IsOrAbove(keybase1.TeamRole_WRITER) {
+		res := keybase1.BoxAuditAttemptResult_OK_NOT_ATTEMPTED_ROLE
+		return false, &res, nil
+	}
+
+	return true, nil, nil
+}
+
+// loadTeamForBoxAudit loads a team once, but if the client
+// has not yet stored BoxSummaryHashes (due to being an old client)
+// it does a force full reload so it is populated.
+func loadTeamForBoxAudit(mctx libkb.MetaContext, teamID keybase1.TeamID) (*Team, error) {
+	return loadTeamForBoxAuditInner(mctx, teamID, false)
+}
+
+func loadTeamForBoxAuditInner(mctx libkb.MetaContext, teamID keybase1.TeamID, force bool) (team *Team, err error) {
+	defer mctx.TraceTimed("loadTeamForBoxAuditInner", func() error { return err })()
+	arg := keybase1.LoadTeamArg{
+		ID:              teamID,
+		ForceRepoll:     true,
+		Public:          teamID.IsPublic(),
+		ForceFullReload: force,
+	}
+
+	team, err = Load(mctx.Ctx(), mctx.G(), arg)
+	if err != nil {
+		return nil, err
+	}
+	if team == nil {
+		return nil, fmt.Errorf("got nil team from loader")
+	}
+
+	// If the team sigchain state was constructed with support for the
+	// merkleRoots map, the map will be non-nil but empty. It will only be nil
+	// if the state is cached from a team load before box summary hash support.
+	if team.chain().GetMerkleRoots() == nil {
+		if force {
+			return nil, fmt.Errorf("failed to get a non-nil merkleRoots map after full reload")
+		}
+		return loadTeamForBoxAuditInner(mctx, teamID, true)
+	}
+	return team, nil
+}
+
+type merkleCheckpoints map[keybase1.UserVersion]keybase1.Seqno
+
+func getAllPUKCheckpointsSinceMerkleSeqno(teamchain *TeamSigChainState, merkleSeqnoCheckpoint keybase1.Seqno) (merkleCheckpoints, error) {
+	checkpoints := make(merkleCheckpoints)
+	// We only check users currently in the team, which means we skip over any
+	// users, who for example, have reset (and possibly have added a new PUK),
+	// but have not been let back into the team by an admin.
+	for uv, logPoints := range teamchain.inner.UserLog {
+		logPoint := logPoints[len(logPoints)-1]
+		if logPoint.Role == keybase1.TeamRole_NONE {
+			continue
+		}
+		merkleSeqno := logPoint.SigMeta.PrevMerkleRootSigned.Seqno
+		latest := merkleSeqno
+		if latest < merkleSeqnoCheckpoint {
+			latest = merkleSeqnoCheckpoint
```

Comment:
this one took me a moment to grasp as well, can we have a comment here that:
> we gather merkle root seqnos for the points in time in which each user has been keyed in the team, which is either the merkle seqno of key rotation, or merkle seqno of change_membership if user has been added afterwards

if that is correct, of course

Reply:
oh but that's only true for the case when we give it merkleSeqnoCheckpoint of key generation. maybe it's fine then

Reply:
if we are making a "current summary" for recently fetched root, we should only see logPoints with root seqno < latest, right? if not, we entered another race